### PR TITLE
Feat/negative ownership

### LINKED ISSUE #63: negative metadata filter on ownership attribute
we currently have the metadata filter on ownership attribute based on a predefined function `contain`.  It would be nice if we have similar behavior but using the function `not contain`. The default value for links not defining the ownership attribute should be an empty dict. In other words, if the link does not define the ownership, it obviously will not match the `not_ownwership` filter.

Refs: https://github.com/kytos-ng/kytos/blob/master/docs/blueprints/EP023-2.rst
Refs: https://github.com/kytos-ng/kytos/blob/master/docs/blueprints/EP023.rst

### PR DESCRIPTION
**Pull request by Ktmi**
Closes #63

### Summary

This PR adds in the ability to apply a filter which rejects paths containing a set of given owners. Additionally, this changes it so that a set of possible owners can be filtered for rather than a single owner.

### Local/e2e Tests

Haven't been able to run extensive local tests yet. I mainly wanted to get this PR into a state where we can start reviewing it.

### CODE REVIEW

File: tests/integration/test_paths_simple.py
Comment (viniarck) [inline tests/integration/test_paths_simple.py:116]: Unit test should still pass with link metadata using `{"ownership": "str"}` value since we do have prod usage using this format, so let's ensure unit test cov and still not break compat.

File: graph/graph.py
Comment (viniarck) [inline graph/graph.py:45]: Same as my previous question, why not just providing a `operator.not_(operator.contains(*, *))` equivalent of `not in` filter? I think it'd be more readable and easier to understand, and then in the future when multiple ownership needs to be filtered as an input then we can get back to this discussion, unless you've already sorted that out with network engineers and documented it. Let me know if you had other plans/ideas in mind.

**Review by viniarck — COMMENTED** [review]
I've just given my initial feedback on this PR. I'm indifferent regarding the new filter class structures, I can see the preprocessor callable also being handy and the type hints are great, so it's great to also have this in place. Now regarding the other changes with ownership logical I have some questions/concerns.

File: tests/integration/test_paths_simple.py
Comment (Ktmi) [reply tests/integration/test_paths_simple.py]: Undid this change, tests now back to there original state

File: graph/graph.py
Comment (Ktmi) [reply graph/graph.py:53]: Did it this way because this is equivalent to the logic that @italovalcy used [here](https://github.com/kytos-ng/pathfinder/issues/63#issuecomment-1819592771).

File: graph/graph.py
Comment (Ktmi) [reply graph/graph.py]: Discussed it with @italovalcy, he agreed that the ambiguity would be a bit confusing. Changed it to `contains`

File: graph/graph.py
Comment (Ktmi) [reply graph/graph.py:56]: Now just uses the raw string rather than splitting it.

File: graph/graph.py
Comment (viniarck) [reply graph/graph.py:52]: I see. Makes sense, and it's also following a similar behavior with `undesired_links` in terms of filtering out any non empty intersection, so it's aligned with the existing multiple values exclusion filters.

File: graph/graph.py
Comment (viniarck) [inline graph/graph.py:56]: @Ktmi, when I was reading this code, I found myself multiple times asking if it was checking the expected type for a value that is coming from the API and will be passed down to the `filter`, or if it was type checking and mapping an edge attribute as you've done with `ProcessEdgeAttribute` and `TypeDifferentiatedProcessor` for `ownership`, the latter is clear, but the former isn't, especially folks who are reading this for the first time or might need to patch or augment in the future.

That said, back then we didn't have any validation on API endpoints, so for the values passed down to the `filter` from the mandatory and flexible metrics now we have a stronger guarantee, so I wonder if at this point in time we can get rid of the additional type checking here to make the code even easier to reason. Can you simplify this part, and only keeping maping the str into a set part? We can still have metadata improperly configured, those type errors with the comparator should still be handled though, which form I followed is already in place.

File: graph/graph.py
Comment (viniarck) [inline graph/graph.py]: @Ktmi, "ownership" accepting multiple and including any intersection still needs to be confirmed if indeed it should behave like a `any` or `all`, currently it's behaving like `any`. So if users want to include `{"mandatory_metrics": {"ownership": "red,blue"}}`, what's the expected behavior? Both ways could make sense, but we need to sort out this requirement with network engineers, and then document in the openapi.yml doc accordingly.

**Review by viniarck — COMMENTED** [review]
I've just done another pass and closed some threads. 

Overall looks good, parts of the typechecking can be simplified, and the `owernship` filter accepting multiple values as input is still leaving margin for multiple interpretations of the expected outcome `any` or `all`, it's up to you if you'll confirm the requirement with Italo and/or Jeronimo, or if that part will be left out for a future PR. At this point, you can start moving on with e2e tests and cover the integration/unit tests. Let me know when you need a final review.

File: graph/graph.py
Comment (Ktmi) [reply graph/graph.py:56]: With the way I have the code written at this point, it would probably be for the best to just dedicate the TypeCheckPreprocessor to type checking, and not modify the val. I was already planning on converting the value preprocessor on `not_ownership` into a type differentiated processor, in order to also allow for lists of strings.

File: graph/graph.py
Comment (Ktmi) [reply graph/graph.py]: Moved ownership to just accept a single, rather than deal with the multiple cases, as discussed with @italovalcy.

Comment (viniarck) [reply graph/graph.py]: Right. Let's close this thread then.

File: graph/graph.py
Comment (viniarck) [inline graph/graph.py:47]: All the usages of `TypeCheckPreprocessor` instances except on `ProcessEdgeAttribute`, at this point in time, are just trying to validate an input that has already been validated [at higher levels call sites via the API endpoint](https://github.com/kytos-ng/pathfinder/blob/master/main.py#L85). So, in practice, we shouldn't need to keep revalidating this business logic part here (back then we had to, but not anymore), otherwise it's just more code to maintain and understand. It's great to maintain the new `EdgeFilter` with the exact same structure and ready to also set a callable preprocessor, but until we really need, to need to start passing these here. Less overhead in terms of code, and as you know, code is read way more times than it's written, so let's prioritize for that, while also evolving the way you're envisioning this part. Let me know if you had any other strong reason to still maintain, otherwise let's remove passing `TypeCheckProcessor` to `EdgeFilter` here and in the other attributes that are type checking against the `value`

File: graph/filters.py
Comment (viniarck) [inline graph/filters.py:106]:
```
❯ curl -s -X POST -H 'Content-type: application/json' http://127.0.0.1:8181/api/kytos/pathfinder/v3/ -d '{"source": "00:00:00:00:00:00:00:01:1", "destination":  "00:00:00:00:00:00:00:03:
1", "spf_max_path_cost": 10, "mandatory_metrics": {"ownership": "yellow"}}' | jq
{
  "description": "Expected types: dict_keys([<class 'str'>, <class 'dict'>, <class 'NoneType'>])",
  "code": 400
}
```

If an edge metadata were to be improperly configured with a type, the current error message is somewhat helpful (especially for folks who understand how it's implemented), but it's not that easy to understand what the error was. Can you try to make this error message with a bit more context just so network operators might be able to easily figure out if it ever happens? Maybe also including the `value` and its type in the error?

**Review by viniarck — CHANGES_REQUESTED** [review]
@Ktmi, appreciated the updates and how the code has greatly evolved. 
I've asked a few minor changes, if you can take a look. 

If you can cover `not_owernship` with unit/integration tests that'd be great. Also, ideally `not_owernship` deserves a e2e since it can be helpful for network engineers to understand the usage, however if unit/integration tests are already properly covering it, then it's OK to defer the e2e test for a subsequent opportunity in that case. Let me know how you're planning to test it.

File: graph/graph.py
Comment (Ktmi) [reply graph/graph.py:47]: I completely forgot about the API validation.  Yeah, I'll get rid of usage of the `TypeCheckPreprocesor`.

File: graph/filters.py
Comment (Ktmi) [reply graph/filters.py]: Added in some more information here. It should now include the specific edge, attribute, and the repr of the value of the attribute.

**Review by viniarck — APPROVED** [review]
@Ktmi, I appreciated the latest updates. I'll leave the PR pre-approved, once unit/integrations tests are covered too feel free to merge.

**Comment by Ktmi** [comment]
@viniarck  I'm going to say merge it at this point. The UI can be a future issue. I've done some local testing with the feature, and everything looks good. I can get mef_eline to dynamically create paths using the `not_ownership` constraint, and have been able to validate that these paths are correct.

**Review by viniarck — APPROVED** [review]
@Ktmi, right. LGTM. However, before merging, based on your last comment:

- Create a new separated issue for `2023.1` just to it can get backported, otherwise the linked issue will get closed too. 
- Create a new issue for the UI part for `2023.2`
- Create a new issue covering e2e test with `not_ownership` for `2023.2`, the current unit tests are great, but let's cover this one on e2e test too

Feel free to merge.

**Pull request merged**
